libbpf-tools/lsan.c

```diff
@@ -0,0 +1,1331 @@
+/* SPDX-License-Identifier: BSD-2-Clause */
+/* Copyright (c) 2022 LG Electronics */
+
+// 19-Jul-2022   Bojun Seo   Created this.
+#include <argp.h>
+#include <stdio.h>
+#include <unistd.h>
+#include <time.h>
+#include <string.h>
+#include <dirent.h>
+#include <stdlib.h>
+#include <sys/uio.h>
+#include <sys/ptrace.h>
+#include <sys/user.h>
+#include <sys/wait.h>
+#include <bpf/libbpf.h>
+#include <bpf/bpf.h>
+#include "lsan.h"
+#include "lsan.skel.h"
+#include "trace_helpers.h"
+#include "uprobe_helpers.h"
+#include "cvector.h"
+#include "uthash.h"
+
+enum log_level {
+	DEBUG,
+	INFO,
+	WARN,
+	ERROR,
+};
+
+void __p(enum log_level level, char *level_str, char *fmt, ...);
+void set_log_level(enum log_level level);
+
+#define p_debug(fmt, ...) __p(DEBUG, "Debug", fmt, ##__VA_ARGS__)
+#define p_info(fmt, ...) __p(INFO, "Info", fmt, ##__VA_ARGS__)
+#define p_warn(fmt, ...) __p(WARN, "Warn", fmt, ##__VA_ARGS__)
+#define p_err(fmt, ...) __p(ERROR, "Error", fmt, ##__VA_ARGS__)
+
+pid_t execute_process(char *);
+
+#define ON_MEM_FAILURE(buf) \
+if (NULL == buf) { \
+	p_err("FATAL: Failed to allocate memory on %s", __func__); \
+	exit(-1); \
+}
+
+/* default */
+static enum log_level log_level = ERROR;
+
+void sig_handler(int signo)
+{
+	if (signo == SIGINT || signo == SIGTERM)
+		kill(0, SIGKILL);
+}
+
+void __p(enum log_level level, char *level_str, char *fmt, ...)
+{
+	va_list ap;
+
+	if (level < log_level)
+		return;
+	va_start(ap, fmt);
+	fprintf(stderr, "%s: ", level_str);
+	vfprintf(stderr, fmt, ap);
+	fprintf(stderr, "\n");
+	va_end(ap);
+	fflush(stderr);
+}
+
+void set_log_level(enum log_level level)
+{
+	log_level = level;
+}
+
+pid_t execute_process(char *cmd)
+{
+	const char *delim = " ";
+	char **argv, *ptr, *filepath;
+	pid_t pid = 0;
+	int i;
+
+	if (cmd == NULL) {
+		p_warn("Invalid argument: command not found");
+		exit(-1);
+	} else
+		p_info("Execute child process: %s", cmd);
+
+	argv = calloc(sizeof(char *), strlen(cmd));
+	if (argv == NULL) {
+		p_err("failed to alloc memory");
+		goto cleanup;
+	}
```

Comment:
Well, it saves `filepath` seperately. So `argv` should be the number of spaces in `cmd`, I guess.
You are right, it is not appropriate to use `strlen(cmd)` for the size of `argv`.(It wastes memory)
But because of the overhead calculating the number of spaces, I'm thinking about to use `strlen(cmd)/2` instead.
Since the size of `argv` cannot exceed `strlen(cmd)/2` logically and the `argv` will be freed soon.